### PR TITLE
Drawing tools: add actions to handle events during mark creation

### LINKED ISSUE #1919: Drawing tasks are broken in mobile Safari
**Package**
lib-classifier

**Describe the bug**
Placing multiple marks does not work in mobile Safari on iOS 14. After placing one mark, I'm unable to place any more, even if I delete my first mark.

**To Reproduce**
Try marking penguins in Penguin Watch.
https://frontend.preview.zooniverse.org/projects/penguintom79/penguin-watch/classify

**Expected behavior**
I should be able to count more than one penguin. 🐧 🐧 🐧 

**Smartphone (please complete the following information):**
 - Device: iPhone X
 - OS: iOS 14
 - Browser Safari
 - Version 14

### PR DESCRIPTION
**Pull request by eatyourgreens**
This extends #1921 to fix #1919 by adding `tool.handlePointerMove(event, mark)` and `tool.handlePointerUp(event, mark)` to the tool model. The pointer is captured on mark creation, and pointer events are handled by the drawing surface until the new mark is finished (https://github.com/zooniverse/front-end-monorepo/issues/1919#issuecomment-737989521). `handlePointerMove` and `handlePointerUp` are called during mark creation and allow tools to implement custom validation of marks. By default, a mark is dragged on pointer move and finished on pointer up, but the transcription line tool overrides these to wait for a second pointer down, which finishes the line. Ellipses might also implement this, to allow for two swipes to create both axes of an ellipse.

I've tested this on browserstack, on a Pixel 4 running Android 11, and it seems to work. I think this needs more thorough testing than #1921, so I've opened it as a separate PR. If you only have time to review a small PR, I'd recommend looking at #1921 first.

Package:
lib-classifier

Closes #1919.

# Review Checklist

## General

- [ ] Are the tests passing locally and on Travis?
- [ ] Is the documentation up to date?

## Components
- [ ] Has a storybook story been created or updated?
- [ ] Is the component accessible? 
  - [ ] Can it be used with a screen reader? [BBC guide to testing with VoiceOver](https://bbc.github.io/accessibility-news-and-you/accessibility-and-testing-with-voiceover-os.html)
  - [ ] Can it be used from the keyboard? [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - [ ] Is it passing accessibility checks in the storybook?

## Apps

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `yarn panic && yarn bootstrap` or `docker-compose up --build` and app works as expected?

## Publishing

- [ ] Is the changelog updated?
- [ ] Are the dependencies updated for apps and libraries that are using the newly published library?

## Post-merging

- [ ] Did the app deploy to https://frontend.preview.zooniverse.org/projects/:project-name/:owner or https://frontend.preview.zooniverse.org/about?
- [ ] Is the new feature working or bug now fixed?
  - [ ] Is there a Talk or blog post written to announce the new feature(s)?
- [ ] Is the design working across browsers (Firefox, Chrome, Edge, Safari) and mobile?
  - [ ] Is this approved by our designer?
- [ ] Is this ready for production deployment?

### CODE REVIEW

File: packages/lib-classifier/src/plugins/drawingTools/experimental/models/tools/TranscriptionLineTool/TranscriptionLineTool.js [hunk 1]
```diff
@@ -7,12 +7,13 @@ const TranscriptionLineTool = types.model('TranscriptionLine', {
...
-      mostRecent.initialDrag(mark)
-      mostRecent.finish()
-      return mostRecent
+    function handlePointerDown (event, mark) {
+      mark.initialDrag(event)
+      mark.finish()
+    }
+
+    function handlePointerUp (event, mark) {
+      return
```
Comment (eatyourgreens) [inline packages/lib-classifier/src/plugins/drawingTools/experimental/models/tools/TranscriptionLineTool/TranscriptionLineTool.js:16]: This works but could maybe check if the mark has a length, and finish if so. We only really want to ignore pointerup if the transcription line mark consists of single point.

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js [hunk 2]
```diff
@@ -53,39 +53,42 @@ function InteractionLayer ({
...
-    activeMark.setSubTaskVisibility(false)
+    createMark(event)
     return false
   }
 
   function onPointerMove (event) {
     if (creating) {
-      const { target, pointerId } = event
-      target.setPointerCapture(pointerId)
       activeMark.initialDrag(convertEvent(event))
```
Comment (eatyourgreens) [inline packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js:92]: Maybe move this into the tool models, so that transcription line tools can ignore dragging?

**Comment by eatyourgreens** [comment]
I think these changes might put the drawing tools API in a better place for handling the polygon and freehand tools.

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js [hunk 3]
```diff
@@ -106,22 +109,29 @@ function InteractionLayer ({
     }
   }
 
+  function onPointerUp(event) {
+    if (creating) {
+      activeTool.handlePointerUp && activeTool.handlePointerUp(convertEvent(event), activeMark)
+      if (activeMark.finished) onFinish(event)
```
[the remark this answers is not in the document]
Comment (shaunanoordin) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js:115]: UPDATE: OK there's a very odd example of how the  error - or rather, a group of potential errors - can trigger.
- If you create a mark, and you have a **dragging action** that causes the **pointer up event to trigger _outside_ the created mark,** then you've just created an Insidiously Wonky Mark. (IWM)
- I call it _insidious_ because you won't know the mark has broken something until you do something specific to follow up on it.

Example:
1. with macOS10 + Chrome on **touchscreen emulation,** go to https://local.zooniverse.org:8080/ (I Fancy Cats)
2. Answer the "How many cats are there?" and "What are they doing?" questions however you want.
3. On the Drawing Task, select the Point ("mark the cat tails") tool.
4. ❗ With your cursor on any cat's tail, **hold the mouse button down** to create a mark on the cat's tail, and IMMEDIATELY move the cursor elsewhere (e.g. the cat's head) and then **let go of the mouse button**
5. You should see that the mark (actually an IWM) stays at the cat's tail, not the cat's head. Nothing else seems amiss.
6. Now, attempt to delete the mark (by clicking on the 'X' next to the mark)
  - Expected: the mark is deleted
  - Actual: the IWM is deleted, but the error message `Uncaught TypeError: Cannot read property 'finish' of undefined` appears in the console.
  - Any attempts to add more marks will fail, with the same error message.

The above is true for `touch-drawing-events`.

As an added bonus, I also tested this behaviour on `master` and the IWM behaves differently. On `master`, deleting the mark will cause no errors to pop up. Instead, attempting to add a new mark results in the error: `Uncaught TypeError: Cannot read property 'initialDrag' of undefined, at onPointerMove`

**Review by shaunanoordin — COMMENTED** [review]
## PR Review(?)

Package: `lib-classifier`

This PR attempts to fix #1919, in which adding marks (for Drawing Tasks) on a touchscreen device kinda sorta breaks the Classifier.

- This PR is a major refactor in how drawing actions create marks/how marks handle events. Code looks clean (with one exception, noted below) so I'm mostly testing on the functionality side of things.
- Functionality looks mostly good - I'm able to (on an emulated touchscreen) place Point-type marks on a Drawing Task. This is particularly important for counting penguins on an iPhone. 🐧 
- ⚠️  There is one exception to this otherwise good review: the IWMs.
  - If you create a mark, and you have a **dragging action** that causes the **pointer up event to trigger _outside_ the created mark**, then you've just created an Insidiously Wonky Mark. (IWM)
  - The IWM will look normal, but will break the app if you attempt to perform an _immediate_ follow up action such as deleting the mark.
  - The IWMs highlight the one concern I might have with the code: the `activeMark.finished` calls in `onPointerUp` (and maybe on `onPointerDown`?) **assumes** that an activeMark _must_ exists whenever the `creating` variable is true. I'm not sure this is the case.
  - Further deets in the inline comments.

Tested on localhost (mostly [I Fancy Cats]'s Point marking tool; I did NOT delve too deeply into the ellipsis tool which is its own can of worms since it relies on drag actions on touchscreen...) with macOS+Chrome on **touchscreen emulation**

### Status

Despite my detailed bug reporting, I'm actually generally positive with the state of this PR. The refactor DOES solve #1919 as described and _will_ allow iPhoners to count penguins.

My only question is, _do you want to try to solve the IWM problem in the context of this PR?_ On one hand, it's a niche problem that's separate from the issues stated in 1919 (I mean, we all know that any drag action on the Classifier is going to be wonky on a touchscreen). On the other, the solution to solving that create-with-drag issue _might involve another refactor,_ and if that's the case it becomes a matter of "well, might as take care of it now." (Do note that the IWM error is subtly different between `master` and this branch - they break at the same place in different ways - so this logically means the refactor is twiddling with the fundamentals of the error somehow.)

I'm currently on the fence, but can easily be convinced to hit Approve if the IWM is clearly considered a future PR's problem.

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js [hunk 4]
```diff
@@ -53,40 +53,43 @@ function InteractionLayer ({
...
 
     if (!activeTool.type) {
       return false;
     }
 
     if (creating) {
-      activeMark = activeTool.handlePointerDown && activeTool.handlePointerDown(convertEvent(event))
-      if (activeMark.finished) setCreating(false)
+      activeTool.handlePointerDown && activeTool.handlePointerDown(convertEvent(event), activeMark)
+      if (activeMark.finished) onFinish(event)
```
[the remark this answers is not in the document]
Comment (eatyourgreens) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js:82]: Yeah, the original code doesn't release pointer capture on finish. That shouldn't be a problem, because the browser should automatically release captured pointers anyway, but I thought it was better to have finished marks handled consistently.

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js [hunk 5]
```diff
@@ -53,40 +53,43 @@ function InteractionLayer ({
...
+    mark.setSubTaskVisibility(false)
+  }
+
   function onPointerDown (event) {
+    const { target, pointerId } = event
+    target.setPointerCapture(pointerId)
     if (disabled || move) {
       return true
     }
-    let activeMark
```
[the remark this answers is not in the document]
Comment (eatyourgreens) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js:60]: Yes, I thought the local variable was confusing. I initially missed that `activeMark` didn't refer to the prop here.

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (eatyourgreens) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js:115]: Yes, something I've noticed on master is that dragging while creating points is broken:
- Expected: the new point mark should follow the pointer and creation should happen on pointer up.
- Actual: the new point doesn't follow the pointer but I do get a new mark wherever pointer up fires.

**Comment by eatyourgreens** [comment]
Rebased, now that #1921 has merged, so that this is only concerned with adding the new event handlers.

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (eatyourgreens) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js:115]: I haven't tried deleting my new points, but it sounds like the delete handling isn't removing `activeMark` from the stored task.

[the remark this answers is not in the document]
Comment (eatyourgreens) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js:115]: I'm trying this out in Firefox. I'm using the point drawing task on the following workflow. Clicking and dragging to create a mark, then deleting the mark, then creating a new mark. I'm not seeing any errors in the console and I am allowed to continue creating new marks. Should I be testing in Chrome?
https://localhost:8080/?project=908&workflow=3370

[the remark this answers is not in the document]
Comment (eatyourgreens) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js:115]: No errors or problems in Chrome either, other than the existing bug on master: moving a mark doesn't re-render the drawing surface. 🤔

[the remark this answers is not in the document]
Comment (eatyourgreens) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js:115]: The code isn't resetting `creating` after removing the active mark in `inactivateMark`. I've added that in, and the errors in touch emulation mode seem to have vanished.

[the remark this answers is not in the document]
Comment (eatyourgreens) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js:121]: I've changed that to use `useEffect`, watching `activeMark` for changes and reacting if it's deleted. It seems to work, testing in Chrome's touch emulator at least. You have to tap twice to create a new point, which is odd, but there are no errors in the console.

[the remark this answers is not in the document]
Comment (shaunanoordin) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js:121]: Woot woot! I can confirm that, as of Commit da773d7, I am NO LONGER able to replicate the IWM error on https://localhost:8080/?project=908&workflow=3370 _nor_ I Fancy Cats, using Chrome on touch emulation (so presumably, any touch device). 👍 

For kicks, I ran tests on 48ec50b, and the IWM error did continue to occur there, so da773d7's addition of `useEffect(function onDeleteMark() { ... }` was indeed the magic bullet.

**Review by shaunanoordin — APPROVED** [review]
## PR Review (Update)

Package: `lib-classifier`

This PR attempts to fix #1919, in which adding marks (for Drawing Tasks) on a touchscreen device kinda sorta breaks the Classifier. Following up from the previous review, this is looking good!

- Latest/current code changes + documentation updates look A-OK
- Main functionality tests look good. I can add and delete Point-type drawing marks on an emulated touchscreen. 👍 
  - On a non-touchscreen settings (i.e. with a mouse) I can add, delete, and move/edit Point-type drawing marks normally. 👍 
- Edge-case tests also look good. I'm unable to replicate the previously reported Insidious Wonky Marks errors. 👌 

Tested on mac OS + Chrome (using standard desktop mode, and with touch screen emulation) on localhost (I Fancy Cats, and Jim's testing project) 

### Status

LGTM! I think this PR is ready to roll.
